File: 1644-lowest-common-ancestor-of-a-binary-tree-ii/1644-lowest-common-ancestor-of-a-binary-tree-ii.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def lowestCommonAncestor(self, root, p, q):
        """
        :type root: TreeNode
        :type p: TreeNode
        :type q: TreeNode
        :rtype: TreeNode
        """
        self.foundP = False
        self.foundQ = False
        
        def helper(node):
            if not node:
                return None
            
            left = helper(node.left)
            right = helper(node.right)
            
            if node == p:
                self.foundP = True
                return node
            
            if node == q:
                self.foundQ = True
                return node
            
            if left and right:
                return node
            
            return left if left else right
        
        lca = helper(root)
        return lca if self.foundP and self.foundQ else None

File: 1644-lowest-common-ancestor-of-a-binary-tree-ii/1644-lowest-common-ancestor-of-a-binary-tree-ii.java
class Solution {
    boolean foundP = false;
    boolean foundQ = false;
    
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        TreeNode lca = helper(root, p, q);
        System.out.print(foundQ);
        return (this.foundP && this.foundQ ? lca : null);
    }
    
    private TreeNode helper(TreeNode node, TreeNode p, TreeNode q){
        if (node == null) return null;
        
        TreeNode left = helper(node.left, p, q);
        TreeNode right = helper(node.right, p, q);
        
        if (node.val == p.val){
            this.foundP = true;
            return node;
        }
        if (node.val == q.val){
            this.foundQ = true;
            return node;
        }
         
        
        if ((left != null) && (right != null)) return node;
        
        return (left != null ? left : right);
    }
}
